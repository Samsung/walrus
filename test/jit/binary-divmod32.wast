(module
(func (export "test1") (result i32 i32 i32 i32 i32) (local i32 i32 i32)
  i32.const 4321
  local.set 0
  i32.const 1234567
  local.set 1
  i32.const -1
  local.set 2

  i32.const 1234567
  local.get 0
  i32.div_s

  local.get 1
  i32.const 5432
  i32.rem_s

  i32.const -1234567
  local.get 0
  i32.div_s

  local.get 1
  local.get 0
  i32.rem_s

  i32.const -0x7fffffff
  local.get 2
  i32.div_s

  (; 285, 1503, 4294967011, 3082, 2147483647 ;)
)

(func (export "test2") (result i32 i32 i32 i32 i32) (local i32 i32 i32)
  i32.const 4321
  local.set 0
  i32.const 1234567
  local.set 1
  i32.const 1
  local.set 2

  i32.const 1234567
  local.get 0
  i32.div_u

  local.get 1
  i32.const 5432
  i32.rem_u

  i32.const 0xff000000
  local.get 0
  i32.div_u

  local.get 1
  local.get 0
  i32.rem_u

  i32.const 0xffffffff
  local.get 2
  i32.div_s

  (; 285, 1503, 990092, 3082, 4294967295 ;)
)

(func (export "test3") (result i32) (local i32)
  i32.const 0
  local.set 0

  i32.const 5
  local.get 0
  i32.div_s
  (; exception ;)
)

(func (export "test4") (result i32) (local i32)
  i32.const -1
  local.set 0

  i32.const 0x80000000
  local.get 0
  i32.div_s
  (; exception ;)
)

(func (export "test5") (result i32) (local i32)
  i32.const 0
  local.set 0

  i32.const 5
  local.get 0
  i32.rem_s
  (; exception ;)
)

(func (export "test6") (result i32) (local i32)
  i32.const -1
  local.set 0

  i32.const 0x80000000
  local.get 0
  i32.rem_s
  (; exception ;)
)

(func (export "test7") (result i32) (local i32)
  i32.const 0
  local.set 0

  i32.const 5
  local.get 0
  i32.div_u
  (; exception ;)
)

(func (export "test8") (result i32) (local i32)
  i32.const 0
  local.set 0

  i32.const 5
  local.get 0
  i32.rem_u
  (; exception ;)
)
)

(assert_return (invoke "test1") (i32.const 285) (i32.const 1503) (i32.const 4294967011) (i32.const 3082) (i32.const 2147483647))
(assert_return (invoke "test2") (i32.const 285) (i32.const 1503) (i32.const 990092) (i32.const 3082) (i32.const 4294967295))
(assert_trap (invoke "test3") "integer divide by zero")
(assert_trap (invoke "test5") "integer divide by zero")
(assert_trap (invoke "test7") "integer divide by zero")
(assert_trap (invoke "test8") "integer divide by zero")

