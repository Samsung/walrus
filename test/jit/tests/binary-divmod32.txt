(module
(func (export "test1") (result i32 i32 i32 i32 i32) (local i32 i32 i32)
  i32.const 4321
  local.set 0
  i32.const 1234567
  local.set 1
  i32.const -1
  local.set 2

  i32.const 1234567
  local.get 0
  i32.div_s

  local.get 1
  i32.const 5432
  i32.rem_s

  i32.const -1234567
  local.get 0
  i32.div_s

  local.get 1
  local.get 0
  i32.rem_s

  i32.const -0x7fffffff
  local.get 2
  i32.div_s

  (; 285, 1503, 4294967011, 3082, 2147483647 ;)
)

(func (export "test2") (result i32 i32 i32 i32 i32) (local i32 i32 i32)
  i32.const 4321
  local.set 0
  i32.const 1234567
  local.set 1
  i32.const 1
  local.set 2

  i32.const 1234567
  local.get 0
  i32.div_u

  local.get 1
  i32.const 5432
  i32.rem_u

  i32.const 0xff000000
  local.get 0
  i32.div_u

  local.get 1
  local.get 0
  i32.rem_u

  i32.const 0xffffffff
  local.get 2
  i32.div_s

  (; 285, 1503, 990092, 3082, 4294967295 ;)
)

(func (export "test3") (result i32) (local i32)
  i32.const 0
  local.set 0

  i32.const 5
  local.get 0
  i32.div_s
  (; exception ;)
)

(func (export "test4") (result i32) (local i32)
  i32.const -1
  local.set 0

  i32.const 0x80000000
  local.get 0
  i32.div_s
  (; exception ;)
)

(func (export "test5") (result i32) (local i32)
  i32.const 0
  local.set 0

  i32.const 5
  local.get 0
  i32.rem_s
  (; exception ;)
)

(func (export "test6") (result i32) (local i32)
  i32.const -1
  local.set 0

  i32.const 0x80000000
  local.get 0
  i32.rem_s
  (; exception ;)
)

(func (export "test7") (result i32) (local i32)
  i32.const 0
  local.set 0

  i32.const 5
  local.get 0
  i32.div_u
  (; exception ;)
)

(func (export "test8") (result i32) (local i32)
  i32.const 0
  local.set 0

  i32.const 5
  local.get 0
  i32.rem_u
  (; exception ;)
)
)
