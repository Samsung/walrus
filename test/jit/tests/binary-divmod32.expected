[[[[[[[  Function   0  ]]]]]]]
Local 0: mapped to: 0 range [1 - 16]
Local 1: mapped to: 1 range [3 - 15]
Local 2: mapped to: 2 range [5 - 19]
Number of locals reduced from 3 to 3
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.set
  Index: 1
  Param[0]: ref:2(0)
4: Opcode: i32.const
  Result[0]: i32
5: Opcode: local.set
  Index: 2
  Param[0]: ref:4(0)
6: Opcode: i32.const
  Result[0]: i32
7: Opcode: local.get
  Index: 0
  Result[0]: i32
8: Opcode: i32.div_s
  Param[0]: ref:6(0)
  Param[1]: ref:7(0)
  Result[0]: i32
9: Opcode: local.get
  Index: 1
  Result[0]: i32
10: Opcode: i32.const
  Result[0]: i32
11: Opcode: i32.rem_s
  Param[0]: ref:9(0)
  Param[1]: ref:10(0)
  Result[0]: i32
12: Opcode: i32.const
  Result[0]: i32
13: Opcode: local.get
  Index: 0
  Result[0]: i32
14: Opcode: i32.div_s
  Param[0]: ref:12(0)
  Param[1]: ref:13(0)
  Result[0]: i32
15: Opcode: local.get
  Index: 1
  Result[0]: i32
16: Opcode: local.get
  Index: 0
  Result[0]: i32
17: Opcode: i32.rem_s
  Param[0]: ref:15(0)
  Param[1]: ref:16(0)
  Result[0]: i32
18: Opcode: i32.const
  Result[0]: i32
19: Opcode: local.get
  Index: 2
  Result[0]: i32
20: Opcode: i32.div_s
  Param[0]: ref:18(0)
  Param[1]: ref:19(0)
  Result[0]: i32
21: Opcode: return
  Param[0]: ref:8(0)
  Param[1]: ref:11(0)
  Param[2]: ref:14(0)
  Param[3]: ref:17(0)
  Param[4]: ref:20(0)
------------------------------
Frame size: 0, Arguments size: 32 (total: 32)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[20]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.set
  Addr: sp[24]
  Param[0]: i32 const:2
4: Opcode: i32.const
  Result[0]: i32 unused
5: Opcode: local.set
  Addr: sp[28]
  Param[0]: i32 const:4
6: Opcode: i32.const
  Result[0]: i32 unused
7: Opcode: local.get
  Addr: sp[20]
  Result[0]: i32 unused
8: Opcode: i32.div_s
  Param[0]: i32 const:6
  Param[1]: i32 sp[20]
  Result[0]: i32 sp[0]
9: Opcode: local.get
  Addr: sp[24]
  Result[0]: i32 unused
10: Opcode: i32.const
  Result[0]: i32 unused
11: Opcode: i32.rem_s
  Param[0]: i32 sp[24]
  Param[1]: i32 const:10
  Result[0]: i32 sp[4]
12: Opcode: i32.const
  Result[0]: i32 unused
13: Opcode: local.get
  Addr: sp[20]
  Result[0]: i32 unused
14: Opcode: i32.div_s
  Param[0]: i32 const:12
  Param[1]: i32 sp[20]
  Result[0]: i32 sp[8]
15: Opcode: local.get
  Addr: sp[24]
  Result[0]: i32 unused
16: Opcode: local.get
  Addr: sp[20]
  Result[0]: i32 unused
17: Opcode: i32.rem_s
  Param[0]: i32 sp[24]
  Param[1]: i32 sp[20]
  Result[0]: i32 sp[12]
18: Opcode: i32.const
  Result[0]: i32 unused
19: Opcode: local.get
  Addr: sp[28]
  Result[0]: i32 unused
20: Opcode: i32.div_s
  Param[0]: i32 const:18
  Param[1]: i32 sp[28]
  Result[0]: i32 sp[16]
21: Opcode: return
  Param[0]: i32 sp[0]
  Param[1]: i32 sp[4]
  Param[2]: i32 sp[8]
  Param[3]: i32 sp[12]
  Param[4]: i32 sp[16]

[[[[[[[  Function   1  ]]]]]]]
Local 0: mapped to: 0 range [1 - 16]
Local 1: mapped to: 1 range [3 - 15]
Local 2: mapped to: 2 range [5 - 19]
Number of locals reduced from 3 to 3
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.set
  Index: 1
  Param[0]: ref:2(0)
4: Opcode: i32.const
  Result[0]: i32
5: Opcode: local.set
  Index: 2
  Param[0]: ref:4(0)
6: Opcode: i32.const
  Result[0]: i32
7: Opcode: local.get
  Index: 0
  Result[0]: i32
8: Opcode: i32.div_u
  Param[0]: ref:6(0)
  Param[1]: ref:7(0)
  Result[0]: i32
9: Opcode: local.get
  Index: 1
  Result[0]: i32
10: Opcode: i32.const
  Result[0]: i32
11: Opcode: i32.rem_u
  Param[0]: ref:9(0)
  Param[1]: ref:10(0)
  Result[0]: i32
12: Opcode: i32.const
  Result[0]: i32
13: Opcode: local.get
  Index: 0
  Result[0]: i32
14: Opcode: i32.div_u
  Param[0]: ref:12(0)
  Param[1]: ref:13(0)
  Result[0]: i32
15: Opcode: local.get
  Index: 1
  Result[0]: i32
16: Opcode: local.get
  Index: 0
  Result[0]: i32
17: Opcode: i32.rem_u
  Param[0]: ref:15(0)
  Param[1]: ref:16(0)
  Result[0]: i32
18: Opcode: i32.const
  Result[0]: i32
19: Opcode: local.get
  Index: 2
  Result[0]: i32
20: Opcode: i32.div_s
  Param[0]: ref:18(0)
  Param[1]: ref:19(0)
  Result[0]: i32
21: Opcode: return
  Param[0]: ref:8(0)
  Param[1]: ref:11(0)
  Param[2]: ref:14(0)
  Param[3]: ref:17(0)
  Param[4]: ref:20(0)
------------------------------
Frame size: 0, Arguments size: 32 (total: 32)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[20]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.set
  Addr: sp[24]
  Param[0]: i32 const:2
4: Opcode: i32.const
  Result[0]: i32 unused
5: Opcode: local.set
  Addr: sp[28]
  Param[0]: i32 const:4
6: Opcode: i32.const
  Result[0]: i32 unused
7: Opcode: local.get
  Addr: sp[20]
  Result[0]: i32 unused
8: Opcode: i32.div_u
  Param[0]: i32 const:6
  Param[1]: i32 sp[20]
  Result[0]: i32 sp[0]
9: Opcode: local.get
  Addr: sp[24]
  Result[0]: i32 unused
10: Opcode: i32.const
  Result[0]: i32 unused
11: Opcode: i32.rem_u
  Param[0]: i32 sp[24]
  Param[1]: i32 const:10
  Result[0]: i32 sp[4]
12: Opcode: i32.const
  Result[0]: i32 unused
13: Opcode: local.get
  Addr: sp[20]
  Result[0]: i32 unused
14: Opcode: i32.div_u
  Param[0]: i32 const:12
  Param[1]: i32 sp[20]
  Result[0]: i32 sp[8]
15: Opcode: local.get
  Addr: sp[24]
  Result[0]: i32 unused
16: Opcode: local.get
  Addr: sp[20]
  Result[0]: i32 unused
17: Opcode: i32.rem_u
  Param[0]: i32 sp[24]
  Param[1]: i32 sp[20]
  Result[0]: i32 sp[12]
18: Opcode: i32.const
  Result[0]: i32 unused
19: Opcode: local.get
  Addr: sp[28]
  Result[0]: i32 unused
20: Opcode: i32.div_s
  Param[0]: i32 const:18
  Param[1]: i32 sp[28]
  Result[0]: i32 sp[16]
21: Opcode: return
  Param[0]: i32 sp[0]
  Param[1]: i32 sp[4]
  Param[2]: i32 sp[8]
  Param[3]: i32 sp[12]
  Param[4]: i32 sp[16]

[[[[[[[  Function   2  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.get
  Index: 0
  Result[0]: i32
4: Opcode: i32.div_s
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i32
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[4]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.get
  Addr: sp[4]
  Result[0]: i32 unused
4: Opcode: i32.div_s
  Param[0]: i32 const:2
  Param[1]: i32 sp[4]
  Result[0]: i32 sp[0]
5: Opcode: return
  Param[0]: i32 sp[0]

[[[[[[[  Function   3  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.get
  Index: 0
  Result[0]: i32
4: Opcode: i32.div_s
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i32
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[4]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.get
  Addr: sp[4]
  Result[0]: i32 unused
4: Opcode: i32.div_s
  Param[0]: i32 const:2
  Param[1]: i32 sp[4]
  Result[0]: i32 sp[0]
5: Opcode: return
  Param[0]: i32 sp[0]

[[[[[[[  Function   4  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.get
  Index: 0
  Result[0]: i32
4: Opcode: i32.rem_s
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i32
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[4]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.get
  Addr: sp[4]
  Result[0]: i32 unused
4: Opcode: i32.rem_s
  Param[0]: i32 const:2
  Param[1]: i32 sp[4]
  Result[0]: i32 sp[0]
5: Opcode: return
  Param[0]: i32 sp[0]

[[[[[[[  Function   5  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.get
  Index: 0
  Result[0]: i32
4: Opcode: i32.rem_s
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i32
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[4]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.get
  Addr: sp[4]
  Result[0]: i32 unused
4: Opcode: i32.rem_s
  Param[0]: i32 const:2
  Param[1]: i32 sp[4]
  Result[0]: i32 sp[0]
5: Opcode: return
  Param[0]: i32 sp[0]

[[[[[[[  Function   6  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.get
  Index: 0
  Result[0]: i32
4: Opcode: i32.div_u
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i32
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[4]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.get
  Addr: sp[4]
  Result[0]: i32 unused
4: Opcode: i32.div_u
  Param[0]: i32 const:2
  Param[1]: i32 sp[4]
  Result[0]: i32 sp[0]
5: Opcode: return
  Param[0]: i32 sp[0]

[[[[[[[  Function   7  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.get
  Index: 0
  Result[0]: i32
4: Opcode: i32.rem_u
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i32
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[4]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.get
  Addr: sp[4]
  Result[0]: i32 unused
4: Opcode: i32.rem_u
  Param[0]: i32 const:2
  Param[1]: i32 sp[4]
  Result[0]: i32 sp[0]
5: Opcode: return
  Param[0]: i32 sp[0]

test1() => i32:285, i32:1503, i32:4294967011, i32:3082, i32:2147483647
test2() => i32:285, i32:1503, i32:990092, i32:3082, i32:4294967295
test3() => i32:0
test4() => i32:0
test5() => i32:0
test6() => i32:0
test7() => i32:0
test8() => i32:0
